File: project/src/main/java/br/com/project/security/JwtAuthorizationFilter.java
package br.com.project.security;

import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Collections;

import static br.com.project.security.SecurityConstants.*;

public class JwtAuthorizationFilter extends BasicAuthenticationFilter {

    public JwtAuthorizationFilter(AuthenticationManager authenticationManager) {
        super(authenticationManager);
    }

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
        String token = request.getHeader(HEADER);

        if(token == null || !token.startsWith(TOKEN_PREFIX)) {
            filterChain.doFilter(request, response);
            return;
        }

        UsernamePasswordAuthenticationToken auth = getAuthentication(token);

        // define o usuário como autenticado no contexto da aplicação
        SecurityContextHolder.getContext().setAuthentication(auth);
        filterChain.doFilter(request, response);
    }

    public UsernamePasswordAuthenticationToken getAuthentication(String token) {

        if(token == null) {
            return null;
        }

        // decodifica o token para recuperar o username
        String username = JWT.require(Algorithm.HMAC512(SECRET.getBytes()))
                .build()
                .verify(token.replace(TOKEN_PREFIX, ""))
                .getSubject();

        if(username == null) {
            return null;
        }

        return new UsernamePasswordAuthenticationToken(
                username, null, Collections.emptyList()
        );
    }
}

File: src/main/kotlin/br/com/project/projetoIntegrador/security/JwtAuthorizationFilter.kt
package br.com.project.projetoIntegrador.security

import br.com.project.projetoIntegrador.security.SecurityConstants.HEADER
import br.com.project.projetoIntegrador.security.SecurityConstants.SECRET
import br.com.project.projetoIntegrador.security.SecurityConstants.TOKEN_PREFIX
import br.com.project.projetoIntegrador.service.UserDetailsService
import br.com.project.projetoIntegrador.service.UserService
import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.apache.tomcat.util.http.parser.Authorization
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter
import java.util.Collections

class JwtAuthorizationFilter(authManager: AuthenticationManager,
    private val service: UserDetailsService
): BasicAuthenticationFilter(authManager) {

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        chain: FilterChain
    ) {
        val token = request.getHeader(HEADER)

        if(token == null || !token.startsWith(TOKEN_PREFIX)) {
            chain.doFilter(request, response)
            return
        }

        val authToken = getAuthentication(token)

        SecurityContextHolder.getContext().authentication = authToken
        chain.doFilter(request, response)
    }

    private fun getAuthentication(token: String): UsernamePasswordAuthenticationToken {
        val email = JWT.require(Algorithm.HMAC512(SECRET.encodeToByteArray()))
            .build()
            .verify(token.replace(TOKEN_PREFIX, ""))
            .subject
        val user = service.loadUserByUsername(email)

        return UsernamePasswordAuthenticationToken(email, null, user.authorities)
    }
}